### ISSUE TITLE
Site to Site Tunnel Routing issues?

### ISSUE THREAD

**Issue by kriszabala** [gitbls/pistrong#5]
I have 2 networks that I'm trying to connect with LAN ranges 192.168.1.0/24 and 192.168.0.0/24. Both networks have an RPi running pistrong and it looks like the tunnel is established correctly. I can ping between the networks from the pi's themselves but no dice when trying from another node on the network which leads me to believe it's a routing issue.

Running `sudo iptables -t nat --list` on each pi gives the following:

```
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination         

Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination         
ACCEPT     all  --  $therange       anywhere             policy match dir out pol ipsec
MASQUERADE  all  --  $therange       anywhere            

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination 
```
$therange is the range of the other end of the tunnel. 

Running `sudo iptables -L -v` on each end gives the following:
From pi @ 192.168.1.10
```
Chain INPUT (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain FORWARD (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         
    3   252 ACCEPT     all  --  eth0   any     192.168.0.0/24       192.168.1.0/24       policy match dir in pol ipsec reqid 1 proto esp
    3   252 ACCEPT     all  --  any    eth0    192.168.1.0/24       192.168.0.0/24       policy match dir out pol ipsec reqid 1 proto esp

Chain OUTPUT (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination        
``` 

From pi @ 192.168.0.10
```
Chain INPUT (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain FORWARD (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         
    0     0 ACCEPT     all  --  eth0   any     192.168.1.0/24       192.168.0.0/24       policy match dir in pol ipsec reqid 1 proto esp
    0     0 ACCEPT     all  --  any    eth0    192.168.0.0/24       192.168.1.0/24       policy match dir out pol ipsec reqid 1 proto esp

Chain OUTPUT (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         
```

I know I'm probably just overlooking something simple. Any ideas why I can't reach the other side of the tunnel from other nodes?

**Comment by gitbls** [comment]
Kris, just to confirm: You're saying that sitting on, for instance, 192.168.0.10 you're unable to ping hosts on the 192.168.1.x network (and vice versa). Is that correct? If that's the case, I'm pretty sure (but my notes are lacking on this) that the .x systems (besides the VPN server) need to know a route to the .other.x network.

What is the result if you add a route to one of these systems that is along the lines of "192.168.other.0/24 goes to the VPN server on my network"?

**Comment by gitbls** [comment]
OK, had a minute to test and validate. Without a route on an "other node" on one side of the network, that "other node" was unable to access hosts across the VPN. I added the equivalent (in your network terms) of 'route add -net 192.168.1.0/24 gw 192.168.0.10' (for the .0.0/24 network). There are many ways to make this permanent, depending on how your network is configured and what the various host OS are. LMK if you have any further questions/issues on this.

**Comment by kriszabala** [comment]
Thanks for the super quick replies on this. Sorry, I may have been confusing with the use of terminology. Here's a quick and dirty diagram. 
<img width="544" alt="site-to-site-tunnel" src="https://user-images.githubusercontent.com/153799/103571868-a4fdee80-4e99-11eb-9dc6-0a73690e4d44.png">

From either of the RPi's I can reach any host on the other end of the tunnel. I cannot, however, reach Host 1 from Host 0 or vice versa. I guess I was under the assumption that the site-to-site tunnel would accomplish this. Apologies in advance, my networking knowledge is very limited.

**Comment by gitbls** [comment]
Yes, my previously mentioned route statement would be done as thus for your network (assumes Linux):

Host 0: route add -net 192.168.1.0/24 gw 192.168.0.10
Host 1: route add -net 192.168.0.0/24 gw 192.168.1.10

**Comment by kriszabala** [comment]
Ah, got it. The hosts are standalone hardware on which I can't configure routes on. The use case is, one host sends a broadcast and the other will reply. I was thinking the RPi would handle this, filtering for any traffic destined for the subnet of the other end of the tunnel and then forwarding it through the tunnel.

**Comment by gitbls** [comment]
How do these devices get an IP address, and what operating environment is running on them? What are you using for a DHCP server on each of these subnets?

**Comment by kriszabala** [comment]
Each of the routers are the DHCP servers for each sides. I have two use cases for this. 

In one use case, one is a standalone hardware device (laser projector fixture w/ unknown OS, embedded device) and communicates with software running on windows. 

In another use case, again one side is another standalone hardware device and communicates with multiplatform software so it could be windows, linux, android, etc.

So I guess I'm trying to accomplish this in the most general way so that different scenarios will work.

**Comment by gitbls** [comment]
Some DHCP servers can push routes to the DHCP clients when an address is assigned. Worth checking to see if yours does, and if so, will your devices honor it. My only other suggestion would be to use another level of indirection, and have the hardware devices connected directly to a Pi that can forward all traffic appropriately out through a 2nd adapter to the VPN server. You might be able to do this with a dedicated adapter on the VPN server, but might be easier to try and debug starting out with another Pi if that's possible. Sorry I don't have a definitive solution for you.

**Comment by kriszabala** [comment]
Thanks for the ideas! I think for my use case, it will be better for me to upgrade to routers with VPN tunnel functionality. I wasn't sure if the pi's would provide enough bandwidth anyways.

**Comment by gitbls** [comment]
It dawned on me when you closed this...You can easily run the RPis as routers by adding a second ethernet (USB dongle) to each of the Pis. Then, your devices will sit on the Pi's private network and the other network will be on the internet, running the VPN.

In fact, makeTunnel generates the iptables for each end needed for this scenario, in the file Tunnel-<tname>-<otherhost>-firewall.iptables. 

I've done a moderate amount of testing with this, so can't guarantee that it's 100% complete, but if you were to go down this path, I'd be happy to work with you on it as needed.

**Comment by kriszabala** [comment]
I actually just got this to work. Here's what I did. I added a static route on each router and each pi.
Router@192.168.0.1 adds route to 192.168.1.0/24 with the pi@192.168.0.10 as the gateway.
On the pi@192.168.0.10, `route add -host 192.168.1.10 dev eth0` and then `route add -net 192.168.1.0/24 gw 192.168.1.10`
Then repeated for symmetry on the other side's router and pi.

So basically the traffic intended for the other side gets routed to the pi gateway, which in turn routes it to pi on the other side.

Seems to be working, I can now reach Host1 from Host0. See any issues with this?
